File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cs
public class Solution {
    public int[] MaxSlidingWindow(int[] nums, int k) {
        int maxi = int.MinValue;
        PriorityQueue<int,int> pq = new PriorityQueue<int,int>();
        List<int> ans = new List<int>();

        for(int i = 0; i < k; i++){
            pq.Enqueue(i, -nums[i]);
        }
        ans.Add(nums[pq.Peek()]);     
        
        for(int i=k;i<nums.Length;i++){
            pq.Enqueue(i,-nums[i]);
            while(pq.Count > 0 && pq.Peek() <= i - k){
                pq.Dequeue();
            }
            ans.Add(nums[pq.Peek()]);      
        }
        return ans.ToArray();
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        priority_queue<pair<int,int>> pq;
        vector<int> v;

        for(int i=0;i<k;i++){
            pq.push({nums[i],i});
        }
        v.push_back(pq.top().first);
        
        for(int i=k;i<nums.size();i++){
            pq.push({nums[i],i});
            while(pq.top().second-i < 1-k) pq.pop();

            v.push_back(pq.top().first);
        }

        return v;
    }
};